cmd/admin-handlers.go

```diff
@@ -424,478 +393,217 @@ func (adminAPI adminAPIHandlers) ClearLocksHandler(w http.ResponseWriter, r *htt
-	// and return success.
-	if isDryRun(vars) {
-		writeSuccessResponseHeadersOnly(w)
-		return
-	}
-
-	//We are able to use HealObject for healing an upload since an
-	//ongoing upload has the same backend representation as an
-	//object.  The 'object' corresponding to a given bucket,
-	//object and uploadID is
-	//.minio.sys/multipart/bucket/object/uploadID.
-	numOfflineDisks, numHealedDisks, err := objLayer.HealObject(minioMetaMultipartBucket, uploadObj)
-	if err != nil {
-		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
-		return
-	}
-
-	jsonBytes, err := json.Marshal(newHealResult(numHealedDisks, numOfflineDisks))
-	if err != nil {
-		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
-		return
-	}
-
-	// Return 200 on success.
-	writeSuccessResponseJSON(w, jsonBytes)
-}
-
-// HealFormatHandler - POST /?heal&dry-run
-// - x-minio-operation = format
-// - bucket and object are both mandatory query parameters
-// Heal a given object, if present.
-func (adminAPI adminAPIHandlers) HealFormatHandler(w http.ResponseWriter, r *http.Request) {
-	// Get current object layer instance.
-	objectAPI := newObjectLayerFn()
-	if objectAPI == nil {
-		writeErrorResponse(w, ErrServerNotInitialized, r.URL)
+		writeErrorResponseJSON(w, ErrServerNotInitialized, r.URL)
 		return
 	}
 
 	// Validate request signature.
 	adminAPIErr := checkRequestAuthType(r, "", "", "")
 	if adminAPIErr != ErrNone {
-		writeErrorResponse(w, adminAPIErr, r.URL)
+		writeErrorResponseJSON(w, adminAPIErr, r.URL)
 		return
 	}
 
-	// Check if this setup is an erasure code backend, since
-	// heal-format is only applicable to single node XL and
-	// distributed XL setup.
+	// Check if this setup has an erasure coded backend.
 	if !globalIsXL {
-		writeErrorResponse(w, ErrNotImplemented, r.URL)
+		writeErrorResponseJSON(w, ErrHealNotImplemented, r.URL)
 		return
 	}
 
-	// if dry-run is set in query-params, return success as
-	// validations are successful so far.
-	vars := r.URL.Query()
-	if isDryRun(vars) {
-		writeSuccessResponseHeadersOnly(w)
+	bucket, objPrefix, hs, clientToken, forceStart, apiErr := extractHealInitParams(r)
+	if apiErr != ErrNone {
+		writeErrorResponseJSON(w, apiErr, r.URL)
 		return
 	}
 
-	// Create a new set of storage instances to heal format.json.
-	bootstrapDisks, err := initStorageDisks(globalEndpoints)
-	if err != nil {
-		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
-		return
+	// Helper function to fetch client address - we use the
+	// X-forwarded-for header if one is present.
+	getClientAddress := func() string {
+		addr := r.RemoteAddr
+		fwdFor := r.Header.Get("X-Forwarded-For")
+		if fwdFor == "" {
+			return addr
+		}
+		return fwdFor
+	}
+
+	type healResp struct {
+		respBytes []byte
+		errCode   APIErrorCode
+		errBody   string
+	}
+
+	// Define a closure to start sending whitespace to client
+	// after 10s unless a response item comes in
+	keepConnLive := func(w http.ResponseWriter, respCh chan healResp) {
```

Comment:
When `keepConnLive` returns, the go-routine started on 554-558 has terminated. Eventually, all go-routines having a reference to `respCh` terminate - this will allow `respCh` to be garbage collected. It is not needed to close the channel.